rclpy/src/rclpy/_rclpy_qos_event.c

```diff
@@ -0,0 +1,334 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcl/event.h>
+
+typedef union _qos_event_callback_data {
+  // Subscription events
+  rmw_requested_deadline_missed_status_t requested_deadline_missed;
+  rmw_liveliness_changed_status_t liveliness_changed;
+  // Publisher events
+  rmw_offered_deadline_missed_status_t offered_deadline_missed;
+  rmw_liveliness_lost_status_t liveliness_lost;
+} _qos_event_callback_data_t;
+
+typedef PyObject * (* _qos_event_data_filler_function)(_qos_event_callback_data_t *);
+
+bool
+_check_rcl_return(rcl_ret_t ret, const char * error_msg)
```

Comment:
nit, `static` on the functions which should only be usable by this file. I think this avoids possible name conflicts.

Reply:
Actually was hoping for your feedback on this pattern - I've `#include "./_rclpy_qos_event.c"` as source, instead of from a header, so the rcply C layer remains a single compilation unit, just separated out these two APIs and their helper functions by file only.

I will definitely mark all the helper functions as static though, don't want their names to leak anywhere.

Reply:
> Actually was hoping for your feedback on this pattern - I've #include "./_rclpy_qos_event.c" as source

Whatever builds is fine with me ;-) . Separate cpython extensions could make it possible to only load extensions if their features are being used, but we woudn't get that savings now because `rclpy.impl` loads all of them anyways.